### PR TITLE
Flow/Javascript support

### PR DESCRIPTION
**Pull request by ajnsit**
This is WIP code for generating Flow (https://flow.org) code from typed-wire (which can be used by javascript). It seems to work but has not been fully tested. The flow types are raw json objects with structure matching that of the json parsing/serialising code generated by typed-wire for other languages. This means that no special json serialiser/parser code generation is needed.

Please review / provide feedback.

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -2,11 +2,14 @@
 {-# LANGUAGE TemplateHaskell   #-}
 module Main where
 
+import Prelude
```
Comment (agrafix) [inline app/Main.hs]: why is this needed?

File: src/TW/Ast.hs [hunk 2]
```diff
@@ -2,6 +2,8 @@
 {-# LANGUAGE OverloadedStrings #-}
 module TW.Ast where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/Ast.hs]: why is this needed?

File: src/TW/BuiltIn.hs [hunk 3]
```diff
@@ -6,6 +6,8 @@ module TW.BuiltIn
     )
 where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/BuiltIn.hs]: why is this needed?

File: src/TW/Check.hs [hunk 4]
```diff
@@ -1,6 +1,8 @@
 {-# LANGUAGE FlexibleContexts #-}
 module TW.Check where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/Check.hs]: why is this needed?

File: src/TW/CodeGen/Elm.hs [hunk 5]
```diff
@@ -5,6 +5,8 @@ module TW.CodeGen.Elm
     )
 where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/CodeGen/Elm.hs]: why is this needed?

File: src/TW/CodeGen/Flow.hs [hunk 6]
```diff
@@ -0,0 +1,121 @@
+{-# LANGUAGE OverloadedStrings #-}
+module TW.CodeGen.Flow
+    ( makeFileName, makeModule
+    , libraryInfo
+    )
+where
+
+import Prelude
```
Comment (agrafix) [inline src/TW/CodeGen/Flow.hs]: why is this needed?

File: src/TW/CodeGen/Flow.hs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+import System.FilePath
+import qualified Data.List as L
+import qualified Data.Text as T
+
+makeFileName :: ModuleName -> FilePath
+makeFileName (ModuleName parts) =
+    (L.foldl' (</>) "" $ map T.unpack parts) ++ ".js"
+
+libraryInfo :: LibraryInfo
+libraryInfo = LibraryInfo "Flow" "flow-typed-wire-utils" "0.1.0.0"
```
Comment (agrafix) [inline src/TW/CodeGen/Flow.hs]: is this an npm library? is it needed for the generated output and where can I find the source?

File: src/TW/CodeGen/Flow.hs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    T.unlines
+    [ "export type " <> fullType <> " ="
+    , "   {| " <> T.intercalate "\n   , " (map makeStructField $ sd_fields sd)
+    , "   |}"
+    , ""
+    ]
+    where
+      fullType =
+          let types = T.intercalate ", " (map unTypeVar $ sd_args sd)
+          in unTypeName (sd_name sd) <> (if length (sd_args sd) > 0 then ("<" <> types <> ">") else types)
```
Comment (agrafix) [inline src/TW/CodeGen/Flow.hs]: simpler: `if null (sd_args sd) then "" else ("<" <> types <> ">")`

File: src/TW/CodeGen/Flow.hs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+makeStructField :: StructField -> T.Text
+makeStructField sf =
+    (unFieldName $ sf_name sf) <> " : " <> (makeType $ sf_type sf)
```
Comment (agrafix) [inline src/TW/CodeGen/Flow.hs]: might we need some escaping here for reserved keywords, etc?

File: src/TW/CodeGen/Flow.hs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+makeEnumDef ed =
+    T.unlines
+    [ "export type " <> fullType
+    , "   = " <> T.intercalate "\n   | " (map makeEnumChoice $ ed_choices ed)
+    , ""
+    ]
+    where
+      fullType =
+          let types = T.intercalate ", " (map unTypeVar $ ed_args ed)
+          in unTypeName (ed_name ed) <> (if length (ed_args ed) > 0 then ("<" <> types <> ">") else types)
```
Comment (agrafix) [inline src/TW/CodeGen/Flow.hs]: This looks like code duplication from above. Can you extract that into a function?

File: src/TW/CodeGen/Haskell.hs [hunk 11]
```diff
@@ -6,6 +6,8 @@ module TW.CodeGen.Haskell
     )
 where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/CodeGen/Haskell.hs]: ?

File: src/TW/CodeGen/PureScript.hs [hunk 12]
```diff
@@ -6,6 +6,8 @@ module TW.CodeGen.PureScript
     )
 where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/CodeGen/PureScript.hs]: ?

File: src/TW/JsonRepr.hs [hunk 13]
```diff
@@ -1,5 +1,7 @@
 module TW.JsonRepr where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/JsonRepr.hs]: ?

File: src/TW/Loader.hs [hunk 14]
```diff
@@ -1,5 +1,7 @@
 module TW.Loader where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/Loader.hs]: ?

File: src/TW/Parser.hs [hunk 15]
```diff
@@ -7,6 +7,8 @@ module TW.Parser
     )
 where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/Parser.hs]: ?

File: src/TW/Types.hs [hunk 16]
```diff
@@ -1,5 +1,7 @@
 module TW.Types where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/Types.hs]: ?

File: src/TW/Utils.hs [hunk 17]
```diff
@@ -1,6 +1,8 @@
 {-# LANGUAGE OverloadedStrings #-}
 module TW.Utils where
 
+import Prelude
```
Comment (agrafix) [inline src/TW/Utils.hs]: ?

**Review by agrafix — CHANGES_REQUESTED** [review]
Thank you very much for your pull request! I've added some comments, what do you think?

File: app/Main.hs (hunk 1 above)
Comment (ajnsit) [reply app/Main.hs]: It's not, I just like having explicit Prelude imports and they snuck in. I'll remove them.

**Comment by agrafix** [comment]
Great, thank you, this looks good :-)

**Pull request merged**